torch/csrc/jit/codegen/cuda/lower_allocation.cpp

```diff
@@ -0,0 +1,365 @@
+    if (init_val != nullptr) {
+      std::vector<kir::IterDomain*> init_dims;
+      for (size_t axis_i = alloc_pos; axis_i < fuser_tv->nDims(); axis_i++) {
+        if (info.buffer->fuserTv()->axis(axis_i)->isReduction()) {
+          continue;
+        }
+        auto ca_id =
+            gpu_lower->lowerValue(fuser_tv->getComputeAtAxis(axis_i).first)
+                ->as<kir::IterDomain>();
+        init_dims.push_back(ca_id);
+      }
+      kir::Expr* init_expr = ir_builder.create<kir::UnaryOp>(
+          UnaryOpType::Set, info.buffer, init_val);
+      for (auto init_loop_it = init_dims.rbegin();
+           init_loop_it != init_dims.rend();
+           ++init_loop_it) {
+        auto id = *init_loop_it;
+        kir::ForLoop* new_loop = nullptr;
+        if (isParallelTypeThread((*init_loop_it)->parallelType())) {
+          std::stringstream ss;
+          ss << id->parallelType();
+          new_loop = ir_builder.create<kir::ForLoop>(
+              ir_builder.create<kir::NamedScalar>(ss.str(), DataType::Int),
+              id,
+              nullptr);
+        } else {
+          new_loop = ir_builder.create<kir::ForLoop>(
+              ir_builder.create<kir::Int>(c10::nullopt), id, nullptr);
+        }
+        init_expr->setParentScope(new_loop);
+        new_loop->body().push_back(init_expr);
+        init_expr = new_loop;
+      }
+      info.init_expr = init_expr;
+    }
+
+    if (is_output)
+      return;
+
+    // Multiply all the dimensions we're going to use for the allocation
+    // together to get the total size
+    kir::Val* size = nullptr;
+    if (alloc_dims.size() == 0) {
+      size = ir_builder.create<kir::Int>(1);
+    } else {
+      size = alloc_dims[0];
+      for (size_t i = 1; i < alloc_dims.size(); i++) {
+        size = ir_builder.mulExpr(size, alloc_dims[i]);
+      }
+    }
+
+    // Create the allocation node
+    info.alloc_expr = ir_builder.create<kir::Allocate>(
+        info.buffer, info.buffer->memoryType(), size);
+  }
+
+  void handle(kir::Expr* expr) {
+    if (!ir_utils::isTVOp(expr) || expr->isA<kir::Allocate>()) {
+      expr->accept(this);
+      return;
+    }
+
+    // // Found where the allocation needs to be inserted
+
+    for (auto out : expr->outputs()) {
+      if (!out->isA<kir::TensorView>()) {
+        continue;
+      }
+
+      auto out_tv = out->as<kir::TensorView>();
+
+      kir::Val* init = nullptr;
+      if (expr->isA<kir::ReductionOp>() && out_tv->fuserTv()->hasReduction()) {
+        init = expr->as<kir::ReductionOp>()->init();
+      }
+
+      bool is_output = std::find(
+                           gpu_lower->kernel()->outputs().begin(),
+                           gpu_lower->kernel()->outputs().end(),
+                           out) != gpu_lower->kernel()->outputs().end();
+
+      // Don't need to alloc outputs, and if we don't need to initialize we're
+      // done.
+      if (is_output && init == nullptr) {
+        return;
```

Comment:
This doesn't matter for now as the number of output tensors is always one, but more strictly, shouldn't this be `continue` instead of `return`?

Reply:
Yes, you're correct. Was thinking of welford work from @shmsong, but uncertain if that will have a kir expr with multiple outputs.

Reply:
I'll change it.

Reply:
Yes welford will be lowered to a kir expr with multiple outputs to make supporting block and grid parallelization easier. Thanks for pointing this out. I can rebase the welford PR once this PR is merged, and try to make necessary modifications if needed.